File: solutions/go/07/meet_in_middle/main.go
package main

import (
	"fmt"
	"sort"
)

func subsetSums(xs []int) []int {
	sums := []int{0}
	for _, x := range xs {
		next := make([]int, 0, len(sums)*2)
		for _, s := range sums {
			next = append(next, s+x)
		}
		sums = append(next, sums...)
	}
	return sums
}

func canMakeSum(xs []int, target int) bool {
	n := len(xs)
	mid := n / 2
	left := subsetSums(xs[:mid])
	sort.Ints(left)
	right := subsetSums(xs[mid:])
	for _, s := range right {
		need := target - s
		i := sort.SearchInts(left, need)
		if i < len(left) && left[i] == need {
			return true
		}
	}
	return false
}

func main() {
	xs := []int{1, 2, 4, 8, 16}
	for _, t := range []int{0, 7, 15, 31, 32} {
		fmt.Println(t, canMakeSum(xs, t))
	}
}

File: solutions/python/07/meet_in_middle/main.py
import bisect


def subset_sums(xs):
    sums = [0]
    for x in xs:
        sums = [s + x for s in sums] + sums
    return sums


def can_make_sum(xs, target):
    n = len(xs)
    mid = n // 2
    left = sorted(subset_sums(xs[:mid]))
    right = subset_sums(xs[mid:])
    for s in right:
        need = target - s
        i = bisect.bisect_left(left, need)
        if i < len(left) and left[i] == need:
            return True
    return False


def main() -> None:
    xs = [1, 2, 4, 8, 16]
    for t in [0, 7, 15, 31, 32]:
        print(t, can_make_sum(xs, t))


if __name__ == "__main__":
    main()
